### ISSUE TITLE
restarting Scylla can take forever if we need to replay commitlog

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#3958]
We have seen many reports that restarting Scylla can take a long time. That never made much sense as to why, since flushing the commitlog should be relatively fast.

I think I now figured out why:

This is our initialization code in main.cc:

```
            if (cl != nullptr) {
                auto paths = cl->get_segments_to_replay();
                if (!paths.empty()) {
                    supervisor::notify("replaying commit log");
                    auto rp = db::commitlog_replayer::create_replayer(qp).get0();
                    rp.recover(paths, db::commitlog::descriptor::FILENAME_PREFIX).get();
                    supervisor::notify("replaying commit log - flushing memtables");
                    db.invoke_on_all([] (database& db) {
                        return db.flush_all_memtables();
                    }).get();
                    supervisor::notify("replaying commit log - removing old commitlog segments");
                    cl->delete_segments(std::move(paths));
                }
            }
```
The boot sequence depends on memtable flushes ending, due to the `.get()` call that blocks it.
We have been always telling ourselves that this *should* be fast, but:

The controller will assign low shares for memtable flushes, because the memory buffer is small - and getting smaller no new writes are coming in, so the controller understands this as "there is no write pressure"

Compactions are already going on at this point, because the compaction manager is started inside the database object.

We call `trigger_compaction()` later, and this is what tricked me before: that doesn't really *start* compactions, so they are for sure already going on.

With the memtable flush shares being low due to the buffers being small, compactions will dominate the startup time.

There are a couple of possible ways to fix it. One of them is bump shares into the controller - this is actually as far as I understand already done in master, just not in released versions. However, first, the bump is not that big because it wasn't meant for a situation like this. Second, if this is blocking startup, then maybe at the very least should give this a very high shares number.

My preferred solution is, therefore, to delay the compaction manager to start until this stage is done.

**Comment by slivne** [comment]
When you restart scylla "in a clean manner" there should not be any commitlogs left on disk. 

We had an issue in the past that we did not clear commitlogs on shutdown. (#2550)

Naturally if scylla is not shutdown cleanly and starts again - replay will be needed.

@glommer can you please clarify when this happens ??

**Comment by glommer** [comment]
When that happens is in the subject: whenever we need to replay the commitlog. Obviously that happens during unclean shutdown, but how does that matter? If we will assume all shutdown are clean we should maybe get rid of the commitlog code =)

Now that I think about it, we have the same issue with drain. We receive a lot of complains about drain taking forever, sometimes up to 45 minutes. The reason for that has to be that at that point there are no new writes so memory used by memtables go lower and lower. If this runs in parallel with anything, it will be dominated and make no progress.

**Comment by slivne** [comment]
It matters - if we have a new bug that causes us to have commitlogs left on clean shutdown, thanks for the clarification.

**Comment by tgrabiec** [comment]
Refs https://github.com/scylladb/scylla/issues/3761

**Comment by glommer** [comment]
@tgrabiec I don't think the solution for 3761 is good enough here.
As I explained, it gives forced flushes an extra bump. But because this is blocking startup I think it should give it a major bump. This directly affects customer experience.

An alternative is just disable compactions at this point.

**Comment by glommer** [comment]
I just checked the code in drain:

it first flushes all memtables with a .get(), and that's why it will take forever. And *then* it stops compactions. At the very least, it should stop compactions first. But there is a comment on drain that says this:
```
                // FIXME: ongoing compaction tasks should be interrupted, not
                // waited for which is what compaction_manager::stop() does now.
```

That explains why drain is so unbearably slow for a lot of our users. What I'd like to see done here is a way to delay compactions from starting and then a way to stop them as the FIXME says.

This will guarantee that both flushes will happen pretty much alone in the system and will be able to progress fast.

**Comment by mailmahee** [comment]
This fix is in master tip currently but needs to go to enterprise asap and even backport.
CC @dorlaor @eyalgutkind

**Comment by glommer** [comment]
@mailmahee please keep enterprise considerations out of scylla open source bug tracker in the future.
